projects/gloo/pkg/plugins/tunneling/plugin.go

```diff
@@ -40,136 +47,175 @@ func (p *plugin) Name() string {
-								return nil, nil, nil, nil, err
-							}
-							inCluster.TransportSocket = &envoy_config_core_v3.TransportSocket{
-								Name:       wellknown.TransportSocketTls,
-								ConfigType: &envoy_config_core_v3.TransportSocket_TypedConfig{TypedConfig: typedConfig},
-							}
-							break
-						}
-					}
-					generatedClusters = append(generatedClusters, generateSelfCluster(selfCluster, selfPipe, originalTransportSocket))
-					forwardingTcpListener, err := generateForwardingTcpListener(cluster, selfPipe, tunnelingHostname, tunnelingHeaders)
-					if err != nil {
-						return nil, nil, nil, nil, err
-					}
-					generatedListeners = append(generatedListeners, forwardingTcpListener)
-					processedClusters.Insert(cluster)
+	reports reporter.ResourceReports,
+) (
+	[]*envoy_config_cluster_v3.Cluster,
+	[]*envoy_config_endpoint_v3.ClusterLoadAssignment,
+	[]*envoy_config_route_v3.RouteConfiguration,
+	[]*envoy_config_listener_v3.Listener,
+	error,
+) {
+	logger := contextutils.LoggerFrom(params.Ctx)
+
+	var newClusters []*envoy_config_cluster_v3.Cluster
+	var newListeners []*envoy_config_listener_v3.Listener
+
+	// track the clusters we have transformed so we don't do it twice
+	processedClusters := sets.Set[string]{}
+
+	// find all upstreams with tunneling enabled
+	for _, us := range params.Snapshot.Upstreams {
+		clusterName := translator.UpstreamToClusterName(us.GetMetadata().Ref())
+
+		// skip if the cluster has already been processed
+		if processedClusters.Has(clusterName) {
+			logger.Warnf("cluster %v already processed", clusterName)
+			continue
+		}
+
+		// regardless of whether we successfully process the upstream, mark it as processed
+		processedClusters.Insert(clusterName)
+
+		newCluster, newListener, warnings, err := processUpstream(params, us, inClusters)
+		// make sure to add warnings to the report before we handle the error
+		if len(warnings) > 0 {
+			for _, warning := range warnings {
+				reports.AddWarnings(us, warning.Error())
+			}
+		}
+
+		if err != nil {
+			reports.AddError(us, err)
+			continue
+		}
+
+		if newCluster != nil || newListener != nil {
+			newClusters = append(newClusters, newCluster)
+			newListeners = append(newListeners, newListener)
+		}
+	}
+
+	return newClusters, nil, nil, newListeners, nil
+}
+
+func processUpstream(
+	params plugins.Params,
+	us *v1.Upstream,
+	inClusters []*envoy_config_cluster_v3.Cluster,
+) (*envoy_config_cluster_v3.Cluster, *envoy_config_listener_v3.Listener, []*translator.Warning, error) {
+	var warningList []*translator.Warning
+
+	// skip if the upstream does not have tunneling enabled
+	httpProxyHostname := us.GetHttpProxyHostname().GetValue()
+	if httpProxyHostname == "" {
+		return nil, nil, nil, nil
+	}
+
+	clusterName := translator.UpstreamToClusterName(us.GetMetadata().Ref())
+
+	// find the cluster to update
+	cluster := findClusters(inClusters, clusterName)
+	if cluster == nil {
+		return nil, nil, nil, fmt.Errorf("cluster %s not found", clusterName)
```

Comment:
What would lead this to happen? What actions should a user take if it does?

Reply:
A cluster would have to not been created earlier in the process. I don't know what would cause it. I assume that a status with details as to why it wasn't not accepted would be on the Upstream. I will update the message to direct users to check the status of the matching Upstream.

Reply:
More context has been added.